domains/runtime/evm/src/fees.rs

```diff
@@ -0,0 +1,98 @@
+use crate::{AccountId, Balance, Balances, OperatorRewards, Runtime, RuntimeCall};
+use codec::Encode;
+use frame_support::traits::{Currency, ExistenceRequirement, Get, Imbalance, WithdrawReasons};
+use pallet_balances::NegativeImbalance;
+use sp_runtime::traits::{DispatchInfoOf, PostDispatchInfoOf, Zero};
+use sp_runtime::transaction_validity::{InvalidTransaction, TransactionValidityError};
+
+pub struct DomainTransactionByteFee;
+
+impl Get<Balance> for DomainTransactionByteFee {
+    fn get() -> Balance {
+        OperatorRewards::domain_transaction_byte_fee()
+    }
+}
+
+pub struct LiquidityInfo {
+    storage_fee: Balance,
+    imbalance: NegativeImbalance<Runtime>,
+}
+
+/// Implementation of [`pallet_transaction_payment::OnChargeTransaction`] that charges transaction
+/// fees and distributes storage/compute fees and tip separately.
+pub struct OnChargeDomainTransaction;
+
+impl pallet_transaction_payment::OnChargeTransaction<Runtime> for OnChargeDomainTransaction {
+    type LiquidityInfo = Option<LiquidityInfo>;
+    type Balance = Balance;
+
+    fn withdraw_fee(
+        who: &AccountId,
+        call: &RuntimeCall,
+        _info: &DispatchInfoOf<RuntimeCall>,
+        fee: Self::Balance,
+        tip: Self::Balance,
+    ) -> Result<Self::LiquidityInfo, TransactionValidityError> {
+        if fee.is_zero() {
+            return Ok(None);
+        }
+
+        let withdraw_reason = if tip.is_zero() {
+            WithdrawReasons::TRANSACTION_PAYMENT
+        } else {
+            WithdrawReasons::TRANSACTION_PAYMENT | WithdrawReasons::TIP
+        };
+
+        let withdraw_result = <Balances as Currency<AccountId>>::withdraw(
+            who,
+            fee,
+            withdraw_reason,
+            ExistenceRequirement::KeepAlive,
+        );
+        let imbalance = withdraw_result.map_err(|_error| InvalidTransaction::Payment)?;
+
+        // Separate storage fee while we have access to the call data structure to calculate it.
+        let storage_fee = DomainTransactionByteFee::get()
```

Comment:
Why are we not collecting the storage fees here? I see this is taken from the consensus chain `transaction_fee` pallet but the logic does not direclty translate here. The fee above, IIUC, would include the compute fee and transaction byte fee on domain runtime state. We still need to include the consensus storage fee here as well

Reply:
>Why are we not collecting the storage fees here? I see this is taken from the consensus chain transaction_fee pallet but the logic does not direclty translate here. 

We still need to pass the storage fee to `correct_and_deposit_fee` in order to get the `compute_fee_and_tip`, I think it is better to collect execution fee and storage fee at the same place.

>The fee above, IIUC, would include the compute fee and transaction byte fee on domain runtime state. We still need to include the consensus storage fee here as well

The `let storage_fee = DomainTransactionByteFee::get() * size` is the consensus storage fee since `DomainTransactionByteFee` is getting from the consensus chain. The fee is charged by the `pallet-transaction-payment` with `LengthToFee` set to `DomainTransactionByteFee`. And you do remind me that we need to include the original `TransactionByteFee` into the `LengthToFee` in order to charge the domain storage fee for the substrate tx.

Reply:
I'm not sure mixing this would be great idea. There is a much simpler approach to this and should also solve the my question here - https://github.com/subspace/subspace/pull/2384/files#r1446264169

There 3 different fees we collect
- Storage fees to compensate operator
- Storage fees for LenghtToFee on domain
- Compute fee

2 & 3 is already handled by the `pallet-transaction-payment` and I dont think we should alter or modify that
We should introduce a new Signed extension for collecting 1st storage fees. This also solves EVM specific OnchargeTransaction if we call this Signed extension before `pre_dispatch_self_contained` call

Reply:
A new Signed extension does not help EVM tx fee because EVM tx doesn't involve any Signed extension at all, they use their own validate and pre_dispatch in `validate_self_contained` and `pre_dispatch_self_contained`.

Sure we can introduce a new Signed extension for just handling 1st storage fees but I prefer the current approach that is similar to the consensus chain and handles all 3 fees at the same place instead of spreading them out in different places.